pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -167,6 +175,66 @@ func (og *operationGenerator) GenerateVolumesAreAttachedFunc(
 	}, nil
 }
 
+func (og *operationGenerator) GenerateBulkVolumeVerifyFunc(
+	pluginNodeVolumes map[types.NodeName][]*volume.Spec,
+	pluginName string,
+	volumeSpecMap map[*volume.Spec]v1.UniqueVolumeName,
+	actualStateOfWorld ActualStateOfWorldAttacherUpdater) (func() error, error) {
+
+	return func() error {
+		attachableVolumePlugin, err :=
+			og.volumePluginMgr.FindAttachablePluginByName(pluginName)
+		if err != nil || attachableVolumePlugin == nil {
+			glog.Errorf(
+				"BulkVerifyVolume.FindAttachablePluginBySpec failed for plugin %q with: %v",
+				pluginName,
+				err)
+			return nil
+		}
+
+		volumeAttacher, newAttacherErr := attachableVolumePlugin.NewAttacher()
+
+		if newAttacherErr != nil {
+			glog.Errorf(
+				"BulkVerifyVolumes failed for getting plugin %q with: %v",
+				attachableVolumePlugin,
+				newAttacherErr)
+			return nil
+		}
+		bulkVolumeVerifier, ok := volumeAttacher.(volume.BulkVolumeVerifier)
+
+		if !ok {
+			glog.Errorf("BulkVerifyVolume failed to type assert attacher %q", bulkVolumeVerifier)
+			return nil
+		}
+
+		attached, bulkAttachErr := bulkVolumeVerifier.BulkVerifyVolumes(pluginNodeVolumes)
+		if bulkAttachErr != nil {
+			glog.Errorf("BulkVerifyVolume.BulkVerifyVolumes Error checking volumes are attached with %v", bulkAttachErr)
+			return nil
+		}
+
+		for nodeName, volumeSpecs := range pluginNodeVolumes {
+			for _, volumeSpec := range volumeSpecs {
+				nodeVolumeSpecs, nodeChecked := attached[nodeName]
+
+				if !nodeChecked {
+					actualStateOfWorld.MarkVolumeAsDetached(volumeSpecMap[volumeSpec], nodeName)
+					continue
+				}
+
+				check := nodeVolumeSpecs[volumeSpec]
```

Comment:
The first check:

```nodeVolumeSpecs, nodeChecked := attached[nodeName]```

Checks if node is there in `attached` map, if it isn't - we mark the volume as detached. We are doing this mostly because we don't rely on the fact that  each node will be present in `attached` map. 

The second check is for - if node is present but volume is marked as detached. It is needed because `attached` map is now of type - `map[type.NodeName][*volume.Spec]bool` .